### PR TITLE
Added deep-embark.lua

### PR DESCRIPTION
**Pull request by AtomicChicken**
I wrote this following a modder's request for a script that enables players to embark underground. It does this by detecting blocks corresponding to the specified depth (any of the three cavern layers or the underworld), finding one that has a sufficient number of valid walkable tiles in it, and teleporting starting units and equipment there. I developed a workaround for triggering the game to properly reveal tiles centered around the target location which involves placing a glowing barrier tiletype and making it disappear; I'm unaware if there's a more straightforward method to do this, but it works quite well. This script requires the user to set up an `onMapLoad.init` file as detailed in the `-help` printout.

### CODE REVIEW

**Comment by clinodev** [comment]
This is very, very cool! I wonder if it couldn't be expanded slightly to make it possible to invoke through the terminal in embarkation rather than solely through the onMapLoad.init file? I also wonder about the lack of a cavern discovery message (for spores,) but I note I also didn't get a magma discovery message and my test embark is able to build magma workshops immediately. I imagine you've tested it.

Regardless, this is a script I would have asked for years ago if I imagined it was possible! 

Thank you!

File: modtools/deep-embark.lua [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+    local oldDO_MEGA = announcementFlags.DO_MEGA
+    announcementFlags.DO_MEGA = false -- turn off popup announcement
+    local oldPAUSE = announcementFlags.PAUSE
+    announcementFlags.PAUSE = false
+    local oldRECENTER = announcementFlags.RECENTER
+    announcementFlags.RECENTER = recenter and true or false
+    local oldA_DISPLAY = announcementFlags.A_DISPLAY
+    announcementFlags.A_DISPLAY = false
+    local oldD_DISPLAY = announcementFlags.D_DISPLAY
+    announcementFlags.D_DISPLAY = recenter and true or false -- won't recenter without an actual announcement
```
Comment (lethosor) [inline modtools/deep-embark.lua]: Looking at https://dfhack.readthedocs.io/en/stable/docs/Lua%20API.html#typed-object-references, I think you could simplify this with:
```lua
local oldFlags = df.global.d_init.announcements.flags.ENDGAME_EVENT_1:new()
...
df.global.d_init.announcements.flags.ENDGAME_EVENT_1:assign(oldFlags)
```

Comment (AtomicChicken) [reply modtools/deep-embark.lua]: A good idea; changed as advised.

**Comment by AtomicChicken** [comment]
I've updated the script with a new arg and a minor fix. It makes reference to changes from DFHack/df-structures#329.

@clinodev, you're welcome! I have indeed tested the script, and can confirm that cavern spores will spread as expected. Regarding your request: I think I'll go ahead and make a separate terminal-friendly wrapper for this script (since modtools scripts aren't meant to be used in this manner).

File: modtools/deep-embark.lua [hunk 2]
```diff
@@ -115,9 +122,46 @@ function getValidEmbarkTiles(block)
...
-function reveal(pos, recenter)
--- creates an unbound glowing barrier at the target location to trigger proper tile revelation when it disappears (fortress mode only)
--- if recenter is true, shifts the view to the target location
+function blockGlowingBarrierAnnouncements(recenter)
+--  temporarily disables the "glowing barrier has disappeared" announcement
+--  announcement settings are restored after 1 tick
+--  setting recenter to true enables recentering of game view to the announcement position
+  local announcementFlags = df.global.d_init.announcements.flags.ENDGAME_EVENT_1 -- glowing barrier disappearance announcement
+  local oldFlags = df.global.d_init.announcements.flags.ENDGAME_EVENT_1:new()
+  oldFlags:assign(announcementFlags) -- backup announcement settings
```
Comment (lethosor) [inline modtools/deep-embark.lua]: I think :new() creates a copy, so you don't need the :assign() call (but I have yet to actually test this)

Comment (AtomicChicken) [reply modtools/deep-embark.lua]: Looks like this is correct; I'll remove it.

**Comment by lethosor** [comment]
Thanks! FYI, I tweaked some things: https://github.com/DFHack/scripts/compare/c9931ad...4be69fb
My general rule of thumb is that modtools is reserved for scripts that users will never run directly - this is a bit of a grey area, because mods definitely *could* use it, but it also works from the console directly (when run at the correct time).

**Comment by AtomicChicken** [comment]
@lethosor 

This script as is **cannot** be run from the console directly, as it is written to perform all its checks and actions immediately upon loading the map (checking for a fortress age of 0 ticks to determine whether it's a fresh embark or not, for instance). Manually running the script from the console within this timeframe is impossible as fortress mode starts off unpaused. Moreover, the lack of qerror messages was intentional, as I visualised an embark location being preset for a particular mod, with the script being silently aborted when the player is not freshly embarking. This change will result in error messages every single time a fortress is reloaded and throughout adventure mode play.

A better approach might be to make a console-friendly front to be run whilst setting up the embark, which waits for the moment of embark to occur before immediately calling this script.

**Comment by lethosor** [comment]
I should clarify that I ran it from the console on the "you have arrived" screen, and it seemed to work there, although I can't guarantee that there weren't any side effects.

**Pull request merged**